src/Init/Lean/Elab/TermBinders.lean

```diff
@@ -119,21 +118,17 @@ else do
 elabBinders #[binder] (fun fvars => x (fvars.get! 1))
 
 @[builtinTermElab «forall»] def elabForall : TermElab :=
-fun stx _ =>
-  -- `forall` binders+ `,` term
-  let binders := (stx.getArg 1).getArgs;
-  let term    := stx.getArg 3;
+fun stx _ => match_syntax stx.val with
+| `(forall $binders*, $term) =>
   elabBinders binders $ fun xs => do
     e ← elabType term;
     mkForall stx.val xs e
+| _ => unreachable!
 
 @[builtinTermElab arrow] def elabArrow : TermElab :=
-fun stx expectedType? => do
-  id ← mkFreshAnonymousIdent stx.val;
-  let dom    := stx.getArg 0;
-  let rng    := stx.getArg 2;
-  let newStx := mkNode `Lean.Parser.Term.forall #[mkAtom "forall", mkNullNode #[mkExplicitBinder id dom], mkAtom ",", rng];
-  elabTerm newStx expectedType?
+adaptExpander $ fun stx => match_syntax stx with
+| `($dom:term -> $rng) => `(forall (a : $dom), $rng)
+| _                    => unreachable!
 
 @[builtinTermElab depArrow] def elabDepArrow : TermElab :=
```

Comment:
We would need a `bracketedBinder` `parser!` after all here to distinguish it from the elaborator above it.

Reply:
Could you please elaborate on this comment?

Reply:
The pattern `$e -> $f` is ambiguous, it could be either an `arrow` or `depArrow` (actually it cannot be parsed as a `depArrow` because it would parse `$e` as an antiquotation for the whole `depArrow` and fail on the remaining input). Analogously to `$dom:term -> $rng` in `elabArrow`, we would need a pattern `$binder:bracketedBinder -> $rng` to refactor `elabDepArrow`. So we should introduce `bracketedBinder` as either a `parser!`, or a pseudo-kind `mkAntiquot "bracketedBinder" none <|> ...`. The pseudo-kind would not introduce an actual new node kind and thus would not check it during matching, which would not matter here, but might in other cases.

Reply:
(We can also just wait until we have a second use for `:bracketedBinder`)